checker-rs/src/notebook_client.rs

```diff
@@ -0,0 +1,351 @@
+        match read_until_slice(&mut self.conn, PROMPT, &mut response).await {
+            Err(e) => {
+                info!("Socket error waiting for user registration {:?}", e);
+                Err(CheckerError::Mumble("Registration failed"))
+            }
+            Ok(0) => {
+                info!("Unexpected EOF waiting for user registration");
+                Err(CheckerError::Mumble("Registration failed"))
+            }
+            Ok(r) => Ok(r),
+        }?;
+
+        if response != b"User successfully registered\n> " {
+            info!("Unexpected response: {}", bytes_debug_repr(&response));
+            return Err(CheckerError::Mumble("Registration Failed"));
+        }
+
+        Ok(())
+    }
+
+    #[instrument("LOGIN")]
+    pub async fn login(&mut self, user: NotebookUser) -> CheckerResult<()> {
+        info!("Logging in as {:?}", &user);
+        self.user = Some(user);
+        let user_ref = self.user.as_ref().unwrap();
+        let conn = &mut self.conn;
+
+        let mut response_buf = Vec::with_capacity(64);
+        async {
+            let login_str = format!("log {} {}\n", user_ref.username, user_ref.password);
+            conn.write_all(login_str.as_bytes()).await?;
+            conn.flush().await?;
+
+            debug!("Waiting for login");
+            read_until_slice(conn, PROMPT, &mut response_buf).await
+        }
+        .await
+        .map_err(|e: std::io::Error| {
+            warn!("Connection-Error on Login: {}", e);
+            CheckerErrorUnfilled::from(e).with_message("Login failed")
+        })?;
+
+        if response_buf != b"Successfully logged in!\n> " {
+            info!("Unexpected response: {}", bytes_debug_repr(&response_buf));
+            return Err(CheckerError::Mumble("Login failed"));
+        }
+
+        Ok(())
+    }
+
+    #[instrument("SET_NOTE")]
+    pub async fn set_note(&mut self, note: &str) -> CheckerResult<()> {
+        info!("Deploying Note {}", note);
+
+        let conn = &mut self.conn;
+
+        let mut response_buf = Vec::with_capacity(64);
+        async {
+            let request = format!("set {}\n", note);
+            debug!("running: '{}'", request);
+            conn.write_all(request.as_bytes()).await?;
+            conn.flush().await?;
+
+            read_until_slice(conn, PROMPT, &mut response_buf).await
+        }
+        .await
+        .map_err(|e| {
+            warn!("Connection-Error: {}", e);
+            CheckerErrorUnfilled::from(e).with_message("Set-Note connection error")
+        })?;
+
+        let response = String::from_utf8(response_buf)?;
+        let note_id = response
+            .split("Note saved! ID is ")
+            .nth(1)
+            .and_then(|substr| substr.split('!').next())
+            .ok_or(CheckerError::Mumble("Failed to set note"))?;
+
+        self.user.as_mut().unwrap().note = Some(note.to_string());
+        self.user.as_mut().unwrap().note_id = Some(note_id.to_string());
```

Comment:
Could be easier to understand the control flow if you would persist to db here directly

Reply:
Disagree, the User now has a note so I give him the neccesary attributes.
Additionally i thought it was sensible to have db-access in the checker-struct, and I didn't want to give additional (basically unneccesary) handles into the client.

Reply:
Just saying it's not obvious that this gets persisted properly when reading through the code. Valid reasons are valid